onnxruntime/core/providers/partitioning_utils.cc

```diff
@@ -0,0 +1,324 @@
+  auto cur_topo_node = new_order.cbegin();
+  auto end_topo_nodes = new_order.cend();
+
+  std::queue<const Node*> nodes_to_process;         // supported nodes to process
+  std::unordered_set<const Node*> processed_nodes;  // supported nodes we have processed
+  std::map<NodeIndex, std::vector<const Node*>> node_groups;
+  std::vector<const Node*> cur_group;
+
+  bool check_excluded_nodes = !excluded_nodes.empty();
+  const auto excluded_nodes_end = excluded_nodes.cend();
+
+  while (cur_topo_node != end_topo_nodes) {
+    const Node* node = *cur_topo_node;
+    ++cur_topo_node;
+
+    if (processed_nodes.find(node) != processed_nodes.cend()) {
+      continue;
+    }
+
+    if (check_excluded_nodes && excluded_nodes.find(node) != excluded_nodes_end) {
+      processed_nodes.insert(node);
+      continue;
+    }
+
+    bool supported = supported_nodes.count(node) != 0;
+    bool in_partition = !cur_group.empty();
+
+    // check if end of a partition.
+    if (in_partition && !supported) {
+#ifndef NDEBUG
+      if (debug_output) {
+        LOGS_DEFAULT(VERBOSE) << "New partition due to " << node_str(*node)
+                              << ". Nodes in old partition: " << cur_group.size() << "\n";
+        LOGS_DEFAULT(VERBOSE) << group_str(cur_group) << "\n";
+      }
+#else
+      ORT_UNUSED_PARAMETER(debug_output);
+#endif
+      node_groups.insert({cur_group.front()->Index(), std::move(cur_group)});
+    }
+
+    // add the node and any connected downstream nodes that we can handle if supported.
+    // if not mark as processed so we know its inputs are available
+    if (supported) {
+      nodes_to_process.push(node);
+    } else {
+      processed_nodes.insert(node);
+    }
+
+    while (!nodes_to_process.empty()) {
+      node = nodes_to_process.front();
+      nodes_to_process.pop();
+
+      if (processed_nodes.find(node) == processed_nodes.cend()) {
+        // add to partition if all inputs available
+        bool inputs_available = true;
+        for (auto cur = node->InputNodesBegin(), end = node->InputNodesEnd(); cur != end; ++cur) {
+          if (processed_nodes.find(&*cur) == processed_nodes.cend()) {
+            inputs_available = false;
+            break;
+          }
+        }
+
+        if (inputs_available) {
+          cur_group.push_back(node);
+          processed_nodes.insert(node);
+
+          for (auto cur = node->OutputNodesBegin(), end = node->OutputNodesEnd(); cur != end; ++cur) {
+            const Node& downstream_node = *cur;
+
+            // nodes will get added to the queue once per input from a supported node.
+            // we need this to happen as they can't be added to the group until all inputs are known to be available.
+            if (supported_nodes.count(&downstream_node) != 0) {
+              nodes_to_process.push(&downstream_node);
+            }
+          }
+        } else {
+          // need another node providing input to be added to the group.
+        }
```

Comment:
Maybe a bit more comments on this one? A node hit here must be a downstream node of cur_topo_node, so it will go through these tests again?